### ISSUE TITLE
$/cancelRequest incorrect error code

### ISSUE THREAD

**Issue by vytautassurvila** [razzmatazz/csharp-language-server#32]
In vscode extension output window receives focus way to often. When it becomes focused then vscode hides terminal window.

At the moment csharp-ls returns error code 32000 for canceled request via `$/cancelRequest`. According to specs it should be 32800. 

https://microsoft.github.io/language-server-protocol/specifications/specification-current/#cancelRequest
> If the request returns an error response on cancellation it is advised to set the error code to ErrorCodes.RequestCancelled.

https://microsoft.github.io/language-server-protocol/specifications/specification-current/#responseMessage
> `export const RequestCancelled: integer = -32800;`

Logs:
```
[LSP   - 8:47:00 AM] {"isLSPMessage":true,"type":"send-request","message":{"jsonrpc":"2.0","id":436,"method":"textDocument/codeAction","params":{"textDocument":{"uri":"file:///home/REDACTED.cs"},"range":{"start":{"line":868,"character":32},"end":{"line":868,"character":37}},"context":{"diagnostics":[{"range":{"start":{"line":868,"character":32},"end":{"line":868,"character":37}},"message":"The name 'order' does not exist in the current context","severity":1,"source":"lsp"}]}}},"timestamp":1650520020837}
[LSP   - 8:47:01 AM] {"isLSPMessage":true,"type":"send-notification","message":{"jsonrpc":"2.0","method":"textDocument/didChange","params":{"textDocument":{"uri":"file:///home/REDACTED.cs","version":199},"contentChanges":[{"range":{"start":{"line":868,"character":32},"end":{"line":868,"character":37}},"rangeLength":5,"text":"shoppingBasket"}]}},"timestamp":1650520021012}
[LSP   - 8:47:01 AM] {"isLSPMessage":true,"type":"send-notification","message":{"jsonrpc":"2.0","method":"$/cancelRequest","params":{"id":436}},"timestamp":1650520021015}
[LSP   - 8:47:01 AM] {"isLSPMessage":true,"type":"receive-response","message":{"jsonrpc":"2.0","id":436,"error":{"data":{"type":"System.Threading.Tasks.TaskCanceledException","message":"A task was canceled.","code":-2146233029},"code":-32000,"message":"A task was canceled."}},"timestamp":1650520021023}
[Error - 8:47:01 AM] Request textDocument/codeAction failed.
  Message: A task was canceled.
  Code: -32000 
[object Object]
```

**Comment by razzmatazz** [comment]
this is great! thanks for investigation, I was having similar problem in emacs but figured out that was because of emacs

**Comment by razzmatazz** [comment]
note for myself: [StreamJsonRpc](https://github.com/microsoft/vs-streamjsonrpc) seems to be doing the right thing here:
- https://github.com/microsoft/vs-streamjsonrpc/blob/6dc57e800961bf40bca23b87c5cf194f9e7c99d7/src/StreamJsonRpc/JsonRpc.cs#L1587
  - https://github.com/microsoft/vs-streamjsonrpc/blob/6dc57e800961bf40bca23b87c5cf194f9e7c99d7/src/StreamJsonRpc/JsonRpc.cs#L1767

note sure why we're getting -32000 and not -32800..

**Comment by razzmatazz** [comment]
> note sure why we're getting -32000 and not -32800..

we have an an answer to this question, -- F# async forms behaves different to C# tasks w.r.t. AggregateException:
- https://github.com/ionide/LanguageServerProtocol/issues/14#issuecomment-1110211628
- https://github.com/dotnet/fsharp/pull/3257
- https://github.com/demystifyfp/FsToolkit.ErrorHandling/issues/154
- https://github.com/fsharp/fslang-suggestions/issues/840

**Comment by razzmatazz** [comment]
this looks to be a solution, to manually unwrap aggregateexception recursively:
- https://github.com/demystifyfp/FsToolkit.ErrorHandling/issues/154#issuecomment-1082382055

**Comment by razzmatazz** [comment]
this should be fixed by 0.5.3, by unwrapping aggregateexception where in C# this is done by TPL automatically but we need some help for F# to unwrap exceptions ourselves:
- https://www.nuget.org/packages/csharp-ls/0.5.3

**Comment by vytautassurvila** [comment]
Thank you! Works great with vscode extension
